src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseSign1Message.cs

```diff
@@ -17,64 +18,108 @@ internal CoseSign1Message(CoseHeaderMap protectedHeader, CoseHeaderMap unprotect
+            try
             {
-                signature = SignWithECDsa((ECDsa)key, toBeSigned, hashAlgorithm);
+                bytesWritten = CreateCoseSign1Message(content, rentedBuffer, key, hashAlgorithm, protectedHeaders, unprotectedHeaders, isDetached, algHeaderValueToSlip, keyType);
+                Debug.Assert(expectedSize == bytesWritten);
+
+                rentedBuffer.AsSpan(0, bytesWritten).CopyTo(destination);
+                return true;
             }
-            else
+            finally
             {
-                signature = SignWithRSA((RSA)key, toBeSigned, hashAlgorithm);
+                ArrayPool<byte>.Shared.Return(rentedBuffer, clearArray: true);
             }
+        }
 
-            return SignCore(encodedProtetedHeaders, CoseHeaderMap.Encode(unprotectedHeaders), signature, content, isDetached);
+        internal static KeyType GetKeyType(AsymmetricAlgorithm key)
+        {
+            return key switch
+            {
+                ECDsa => KeyType.ECDsa,
+                RSA => KeyType.RSA,
+                _ => throw new CryptographicException(SR.Format(SR.Sign1SignUnsupportedKey, key.GetType()))
+            };
         }
 
-        private static byte[] SignCore(
-            ReadOnlySpan<byte> encodedProtectedHeader,
-            ReadOnlySpan<byte> encodedUnprotectedHeader,
-            ReadOnlySpan<byte> signature,
-            ReadOnlySpan<byte> content,
-            bool isDetached)
+        internal static void ValidateBeforeSign(
+            CoseHeaderMap? protectedHeaders, CoseHeaderMap? unprotectedHeaders, KeyType keyType, HashAlgorithmName hashAlgorithm,
+            out int? algHeaderValueToSlip)
+        {
+            ThrowIfDuplicateLabels(protectedHeaders, unprotectedHeaders);
+            algHeaderValueToSlip = ValidateOrSlipAlgorithmHeader(protectedHeaders, unprotectedHeaders, keyType, hashAlgorithm);
+        }
+
+        [UnsupportedOSPlatform("browser")]
+        internal static int CreateCoseSign1Message(
+            ReadOnlySpan<byte> content, Span<byte> buffer,
+            AsymmetricAlgorithm key, HashAlgorithmName hashAlgorithm,
+            CoseHeaderMap? protectedHeaders, CoseHeaderMap? unprotectedHeaders, bool isDetached, int? algHeaderValueToSlip, KeyType keyType)
         {
             var writer = new CborWriter();
             writer.WriteTag(Sign1Tag);
             writer.WriteStartArray(Sign1ArrayLegth);
-            writer.WriteByteString(encodedProtectedHeader);
-            writer.WriteEncodedValue(encodedUnprotectedHeader);
+
+            int protectedMapBytesWritten = CoseHeaderMap.Encode(protectedHeaders, buffer, true, algHeaderValueToSlip);
+            ReadOnlySpan<byte> encodedProtectedHeaders = buffer.Slice(0, protectedMapBytesWritten);
+            writer.WriteByteString(encodedProtectedHeaders);
+
+            int unprotectedMapBytesWritten = CoseHeaderMap.Encode(unprotectedHeaders, buffer.Slice(protectedMapBytesWritten));
```

Comment:
Why slice the destination?  `buffer[0..protectedMapBytesWritten]` has already been read, and isn't needed any more.  Right? (If we're leaving it there because of how the to-be-signed works, then add a comment)

Reply:
yes, we still need the encoded protected headers to create the toBeSigned. See https://datatracker.ietf.org/doc/html/draft-ietf-cose-msg#section-4.4 bullet 2 - `body_protected`.
```cddl
   Sig_structure = [
       context : "Signature" / "Signature1" / "CounterSignature",
       body_protected : empty_or_serialized_map,
       ? sign_protected : empty_or_serialized_map,
       external_aad : bstr,
       payload : bstr
   ]
```

Reply:
Please add that information as a comment in the code.  Something like

```suggestion
            // We're going to use the encoded protected headers again after this step (for the TBS construction),
            // so don't overwrite them yet.
            int unprotectedMapBytesWritten = CoseHeaderMap.Encode(unprotectedHeaders, buffer.Slice(protectedMapBytesWritten));
```